src/main/java/org/folio/rest/impl/CopycatAPI.java

```diff
@@ -4,25 +4,158 @@
 import io.vertx.core.Context;
 import io.vertx.core.Future;
 import io.vertx.core.Handler;
+import io.vertx.core.json.JsonObject;
+import javax.ws.rs.core.Response;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.folio.copycat.JsonMarc;
+import org.folio.copycat.RecordImporter;
 import org.folio.rest.annotations.Validate;
 import org.folio.rest.jaxrs.model.CopyCatImports;
 import org.folio.rest.jaxrs.model.CopyCatTargetCollection;
 import org.folio.rest.jaxrs.model.CopyCatTargetProfile;
+import org.folio.rest.jaxrs.model.Error;
+import org.folio.rest.jaxrs.model.Errors;
+import org.folio.rest.jaxrs.model.TargetOptions;
 import org.folio.rest.jaxrs.resource.Copycat;
 import org.folio.rest.persist.PgUtil;
-import javax.ws.rs.core.Response;
-import java.util.Map;
+import org.folio.rest.persist.PostgresClient;
+import org.yaz4j.Connection;
+import org.yaz4j.PrefixQuery;
+import org.yaz4j.Query;
+import org.yaz4j.Record;
+import org.yaz4j.ResultSet;
+import org.yaz4j.exception.ZoomException;
 
 public class CopycatAPI implements Copycat {
 
+  static Errors createErrors(String message) {
+    List<Error> errors = new LinkedList<>();
+    errors.add(new Error().withMessage(message));
+    return new Errors().withErrors(errors);
+  }
+
+  static Errors createErrors(Throwable throwable) {
+    log.warn(throwable.getMessage(), throwable);
+    return createErrors(throwable.getMessage());
+  }
+
   private static final String PROFILE_TABLE = "targetprofile";
+  private static Logger log = LogManager.getLogger(CopycatAPI.class);
+
+  /**
+   * Construct yaz4j Query based on external identifier and query mappping.
+   * @param profile
+   * @param externalId
+   * @return Query
+   */
+  static Query constructQuery(CopyCatTargetProfile profile, String externalId) {
+    // assuming the externalId does not have whitespace or include {}"\\ characters
+    String pqf = profile.getExternalIdQueryMap().replace("$identifier", externalId);
+    return new PrefixQuery(pqf);
+  }
+
+  static void setAuthOptions(Connection conn, String auth) {
+    if (auth != null) {
+      String[] s = auth.trim().split("\\s+");
+      conn.option("user", s[0]);
+      if (s.length == 2) {
+        conn.option("password", s[1]);
+      } else if (s.length == 3) {
+        conn.option("group", s[1]);
+        conn.option("password", s[2]);
+      }
+    }
+  }
+  /**
+   * Search for identifier and return record.
+   * @param profile Target Profile
+   * @param externalId
+   * @param type render type
+   * @return record content
+   */
+  static Future<byte[]> getMARC(CopyCatTargetProfile profile, String externalId, String type) {
```

Comment:
Maybe refactor to a seperate class, MarcRetriever

Reply:
Good idea. Will do.